vendor/gopkg.in/src-d/go-git.v4/remote.go

```diff
@@ -511,8 +512,29 @@ func (r *Remote) references() ([]*plumbing.Reference, error) {
 	return localRefs, nil
 }
 
-func getHaves(localRefs []*plumbing.Reference) ([]plumbing.Hash, error) {
+func getHaves(
+	localRefs []*plumbing.Reference,
+	remoteRefs storer.ReferenceStorer,
+	s storage.Storer,
+) ([]plumbing.Hash, error) {
 	haves := map[plumbing.Hash]bool{}
+
+	// Build a map of all the remote references, to avoid loading too
+	// many parent commits for references we know don't need to be
+	// transferred.
+	stopEarlyRefs := map[plumbing.Hash]bool{}
+	iter, err := remoteRefs.IterReferences()
+	if err != nil {
+		return nil, err
+	}
+	err = iter.ForEach(func(ref *plumbing.Reference) error {
+		stopEarlyRefs[ref.Hash()] = true
```

Comment:
Can we cache this? Seems like we'll be rebuilding it on every push.

Reply:
Every push is a separate invocation of `git-remote-keybase` and `getHaves` should only be called once for each push, so I'm not sure where we'd cache it.  To disk maybe?  But note that the thing we're iterating over here is entirely in memory -- it just represents an answer from earlier when we asked the remote side to list all its references.  This is just a weird iterator type go-git puts in front of that.

Reply:
All right let's leave it.